tutorial/ndarrays.md

```diff
@@ -0,0 +1,328 @@
+[read more](https://numpy.org/doc/stable/reference/arrays.html)
+
+## Pyccel ndarrays ##
+
+Pyccel uses the same implementation as Numpy ndarrays with some rules due to the difference between the host language (Python) "dynamically typed / internal garbage collector" and the target languages such as C and Fortran which are statically typed languages and don't have a garbage collector.
+
+Below we will show some rules that Pyccel has set to handles those differences.
+
+### Dynamically and statically typed languages ###
+
+Generally a variable in Pyccel should always keep its initial type, this also transfers to using the ndarrays.
+
+#### incorrect example ####
+
+```Python
+import numpy as np
+
+a = np.array([1, 2, 3], dtype=float)
+#(some code...)
+a = np.array([1, 2, 3], dtype=int)
+```
+
+_OUTPUT_ :
+
+```Shell
+ERROR at annotation (semantic) stage
+pyccel:
+ |error [semantic]: ex.py [5]| Incompatible types in assignment (|a| real <-> int)
+```
+
+### Memory management ###
+
+Pyccel makes a difference between ndarrays that own their data and the ones they don't.
+
+Pyccel calls its own garbage collector when needed, but has a set of rules to do so:
+
+-   Can not reassign ndarrays with different ranks.
+
+    ```Python
+    import numpy as np
+
+    a = np.ones((10, 20))
+    #(some code...)
+    a = np.ones(10)
+    ```
+
+    _OUTPUT_ :
+
+    ```Shell
+    ERROR at annotation (semantic) stage
+    pyccel:
+     |error [semantic]: ex.py [4]| Incompatible redefinition (|a| real(10, 20) <-> real(10,))
+    ```
+
+    This limitation is due to the way Fortran allocatable can't change the rank after declaration.
+
+-   Can not assign ndarrays that own their data one another.
+
+    ```Python
+    import numpy as np
+
+    a = np.array([1, 2, 3, 4, 5])
+    b = np.array([1, 2, 3])
+    a = b
+    ```
+
+    _OUTPUT_ :
+
+    ```Shell
+    ERROR at annotation (semantic) stage
+    pyccel:
+     |error [semantic]: ex.py [5]| Arrays which own their data cannot become views on other arrays (a)
+    ```
+
+    This limitation is due to the fact that the ndarray **a** will have to go from a data owner to a pointer to the **b** ndarray data.
+
+    _NOTE_: this limitation is not applied to assignments which reserve a new memory block.
+
+    -   Python example:
+
+        ```Python
+        import numpy as np
+        a = np.ones(20)
+        #(some code...)
+        a = np.ones(10)
+        ```
+
+    -   C equivalent:
+
+        ```C
+        #include <ndarrays.h>
+        #include <stdlib.h>
+        int main()
+        {
+            t_ndarray a;
+            a = array_create(1, (int32_t[]){20}, nd_double);
+            array_fill((double)1.0, a);
+            /*(some code...)*/
+            if (a.shape != NULL)
+            {
+                free_array(a); //this line handles the redefinition of the array
+            }
+            a = array_create(1, (int32_t[]){10}, nd_double);
```

Comment:
Shouldn't we reallocate the array only if its shape changes? Here we only check that `a.shape != NULL`

Reply:
yes, we are reallocating the array only when the shape is changed the `if` condition for the shape is just a safeguard for C.
**it can be removed.**